### DISCUSSION TITLE
Welcome to ZXing.Net Discussions!

### DISCUSSION THREAD

**Discussion by micjahn** [micjahn/ZXing.Net#295] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by manishabb2011** [comment]
Hello,

I am a software developer in medical industry. My project already uses ZXing.net for barcode creation.

I am exploring about setting the DataMatrix module formats/layout e.g. 12x26 or 12x36 in c#.
In fact there is specific requirement to set format as 12x26 with each element size 20mil. 
But I could not find anything on google search.

Please guide.
Thanks

**Reply by micjahn** [reply]
This snippet should work and should set the dimension constraints to your needs.
```
                var writer = new BarcodeWriter
                {
                    Format = BarcodeFormat.DATA_MATRIX,
                    Options = new ZXing.Datamatrix.DatamatrixEncodingOptions
                    {
                        MinSize = new Dimension(26, 12),
                        MaxSize = new Dimension(26, 12),
                    }
                };
                var image = writer.Write("<your content here>");
```

The element size of 20mil depends on your output device. In my opinion you have to stretch the resulting image to your desired size.

**Comment by fikri1985** [comment]
Hello, 

I am VB 2019 programmer. 

I am using ZXing 0.16.6 to create QR, DataMatrix and Code128 barcode as picture at windows GUI.

The source code as below. I have 2 concerns :
1) any way to remove white background at outer side of DataMatrix barcode
2) any way to remove string for Code128 barcode. I used command .PureBarcode = True however string still appear

`	
        Dim QR As QrCodeEncodingOptions
	Dim QRWriter As IBarcodeWriter
	QR = New QrCodeEncodingOptions

	Dim DM As DatamatrixEncodingOptions
	Dim DMWriter As IBarcodeWriter
	DM = New DatamatrixEncodingOptions

	Dim BLLinear As Code128EncodingOptions
	Dim BLLinearWriter As IBarcodeWriter
	BLLinear = New Code128EncodingOptions

	QR.CharacterSet = "UTF-8"
	QR.DisableECI = True
	QR.PureBarcode = True
	QRWriter = New BarcodeWriter
	QRWriter.Format = BarcodeFormat.QR_CODE
	QRWriter.Options.Width = 60
	QRWriter.Options.Height = 60
	QRWriter.Options.Margin = 0
	PictureBox2.Image = New Bitmap(QRWriter.Write("AAbbCC1234567890_DD-MMM-YYY_00,11,22"))

	DM.PureBarcode = True
	DMWriter = New BarcodeWriter
	DMWriter.Format = BarcodeFormat.DATA_MATRIX
	DMWriter.Options.Width = 60
	DMWriter.Options.Height = 60
	DMWriter.Options.Margin = 0
	PictureBox3.Image = New Bitmap(DMWriter.Write("AAbbCC1234567890_DD-MMM-YYY_00,11,22"))

	BLLinear.PureBarcode = True
	BLLinearWriter = New BarcodeWriter
	BLLinearWriter.Format = BarcodeFormat.CODE_128
	PictureBox4.Image = New Bitmap(BLLinearWriter.Write("12345678901234567890"))`

**Reply by micjahn** [reply]
Generate Datamatrix with width and height as 0 to get the smallest possible image without the margin. Scale it manually to your needs.
```
DMWriter = New BarcodeWriter
DMWriter.Format = BarcodeFormat.DATA_MATRIX
DMWriter.Options.Width = 0
DMWriter.Options.Height = 0
DMWriter.Options.Margin = 0
PictureBox3.Image = New Bitmap(DMWriter.Write("AAbbCC1234567890_DD-MMM-YYY_00,11,22"))
```

Set the value for PureBarcode at the options property of the BLLinearWriter
```
BLLinearWriter = New BarcodeWriter
BLLinearWriter.Format = BarcodeFormat.CODE_128
BLLinearWriter.Options.PureBarcode = True
PictureBox4.Image = New Bitmap(BLLinearWriter.Write("12345678901234567890"))
```

**Comment by rmrudro** [comment]
Hello Everyone I want to know how I can implement Concatenated QR Code using Zxing . I work in a Japanease company . I implemented QR code using Zxing I can read and Write QR Code . But I have a question How can I implement Concatenated QR Code. First I want to say what is Concatenated QR Code is .
[Supplemental Information on Concatenated QR Code]

Concatenated QR code is not a function to "divide a character string into multiple symbols".
It is a function to "express one character string with multiple symbols".

To read the symbol output as a concatenated QR code, a reader that supports the concatenation function is required.
Since each symbol also contains information such as "what number out of what", the symbol alone cannot read the correct character string.

If you set double-byte characters, single-character data may be output across two symbols. This is by design.

Many QR code reader apps that can be used on smartphones and tablets do not support the link function.
To check the output of the linked QR code, install an app that announces support for the linked function, or use a dedicated scanner.

Concatenated QR code is written 連結QRコード. Here is the Ruby site It is also written what is Linked or Concatenated QR code.
You can also check it from here https://linuxtut.com/ruby-generate-concatenated-qr-code-with-rqrcode-(knowledge)-3dd4c/
Here it is written What is a linked QR code?
The QR code has a "concatenation function" that divides long data into up to 16 QR codes and combines them when reading. (For details, refer to QR Code.com)
In Japan, you will often see prescriptions from hospitals / clinics.
The output function is supported by third-party form output libraries, but it is difficult to find it in the OSS library, so I have been using it rqrcode. I modified rqrcode) so that it can be output.
This time, the target is to output the concatenated QR code to be recorded in the above "prescription", so only the QR code in 8bit_byte mode is targeted.
In this Japanease site there is also implemented Concatenated QR Code
https://programresource.net/2013/05/04/2193.html
But is written in Java Can you help me how to implement Concatenated QR Code

**Comment by kernalpanic123** [comment]
Hi,
A customer of mine is requesting a GS1 QR code in navision.  Just by inserting the line 
` barcodeWriter.Options.GS1Format := TRUE`
 didnt give me the correct output.  Do i need to do something more so that the brackets are eliminated during scanning and so that the  QR code contain FNC1 Characters  

Thank you in advance.

https://github.com/micjahn/ZXing.Net/releases/download/v0.16.4.0/ZXing.Net.0.16.4.0.zip
```

[TryFunction] GenerateBarcode(content : Text;barcodeType : Text;height : Integer;width : Integer;pureBarcode : Boolean;margin : Integer;rotateDegree : '0,90,180,270';VAR tempBlobVar : TEMPORARY Record TempBlob)
IF NOT tempBlobVar.ISTEMPORARY THEN
  ERROR('tempBlobVar must be temporary');

CASE UPPERCASE(barcodeType) OF
  'ITF': barcodeFormat := barcodeFormat.ITF;
  'CODE_128': barcodeFormat := barcodeFormat.CODE_128;
  //'UPC_A': barcodeFormat := barcodeFormat.UPC_A;
  //'EAN_8': barcodeFormat := barcodeFormat.EAN_8;
  'EAN_13': barcodeFormat := barcodeFormat.EAN_13;
  //'CODE_39': barcodeFormat := barcodeFormat.CODE_39;
  //'CODABAR': barcodeFormat := barcodeFormat.CODABAR;
  //'PLESSEY': barcodeFormat := barcodeFormat.PLESSEY;
  //'MSI': barcodeFormat := barcodeFormat.MSI;
  'QR_CODE': barcodeFormat := barcodeFormat.QR_CODE;
  //'PDF_417': barcodeFormat := barcodeFormat.PDF_417;
  //'AZTEC': barcodeFormat := barcodeFormat.AZTEC;
  //'DATA_MATRIX': barcodeFormat := barcodeFormat.DATA_MATRIX;
  ELSE
    ERROR('Unknow barcodeType: %1', barcodeType);
END;

encodingOptions := encodingOptions.EncodingOptions();
encodingOptions.Height := height;
encodingOptions.Width := width;
encodingOptions.PureBarcode := pureBarcode;
encodingOptions.Margin := margin;

barcodeWriter := barcodeWriter.BarcodeWriter();
barcodeWriter.Format := barcodeFormat;
barcodeWriter.Options := encodingOptions;
//barcodeWriter.Options.GS1Format := TRUE;  
bitmap := barcodeWriter.Write(content);

IF rotateDegree > rotateDegree::"0" THEN BEGIN
  CASE rotateDegree OF
    rotateDegree::"90": rotateFlipType := rotateFlipType.Rotate90FlipNone;
    rotateDegree::"180": rotateFlipType := rotateFlipType.Rotate180FlipNone;
    rotateDegree::"270": rotateFlipType := rotateFlipType.Rotate270FlipNone;
  END;
  bitmap.RotateFlip(rotateFlipType);
END;

imageFormat := imageFormat.Jpeg;
tempBlobVar.Blob.CREATEOUTSTREAM(outstr);
bitmap.Save(outstr, imageFormat);

bitmap.Dispose;
```

**Reply by micjahn** [reply]
barcodeWriter.Options.GS1Format only adds the FNC1 character at the beginning of the mode header.
Groups have to be added by special characters in the content.
Please take a look at the following wiki article: https://github.com/micjahn/ZXing.Net/wiki/Generating-GS1-DataMatrix-and-GS1-Code128-barcodes

**Reply by DBizcochoVitro** [reply]
Hello, I'm experiencing the same issue. I'm using `//barcodeWriter.Options.GS1Format := TRUE;` but I can't get the DataMatrix reading to recognize it as GS1. Additionally, I'm unsure how to input the Char29 as they suggest. Did you find a solution? I can't find anything related to it. Thank you very much.

**Reply by DBizcochoVitro** [reply]
> Hello, I'm experiencing the same issue. I'm using //barcodeWriter.Options.GS1Format := TRUE; but I can't get the DataMatrix reading to recognize it as GS1. Additionally, I'm unsure how to input the Char29 as they suggest. Did you find a solution? I can't find anything related to it. Thank you very much.

**Comment by droselloc** [comment]
Hello, I am using zxing for the development of an application in vb.net winform for the laboratory. I need to generate codes in datamatrix format but I have a problem. It depends on the length of the string to encoder the code comes out square or rectangle. Could it be forced to always be square?

**Reply by micjahn** [reply]
Sample for C#, should be similar in VB.Net:
```
var writer = new BarcodeWriter
{
    Options = new ZXing.Datamatrix.DatamatrixEncodingOptions
   {
      SymbolShape = ZXing.Datamatrix.Encoder.SymbolShapeHint.FORCE_SQUARE
   }
};
var image = writer.Write("<your content here>");
```

**Comment by zBugi** [comment]
Hello, I am still new to C# and I use zxing for <TargetFramework>net5.0</TargetFramework> but it is giving me a Warning. But it works so freaking fast compared to the other libs and didn't have any issue with that should I be aware of that or just ignore this warning, because I don't get any kind of warning message just the yellow annoying triangle on Dependencies ->Packages-> ZXing.Net  in Visual Studio 2019?

**Reply by micjahn** [reply]
Would be great to know what the message of the warning is. Should be visible if you hover over the warning symbol or perhaps it is shown in the output window.

**Comment by mickagau** [comment]
Hello
First of all thank you for this great framework. I learn a lot every time i use it :)

I have a particular need where i would like to set a fix size for a module of pdf417 instead of letting the library to define the best module size in pix. For instance a size of module of 7 pixels in width.

Please could you list where i would have to modify source code to do so (or if not doable at all) ?

Idea would be to send a parameter as a hint (if this hint does not exist then library define best module size in pix as it already does).

Thank you
Mickael

**Reply by micjahn** [reply]
I'm not sure but I think you should take a look at the class ZXing.PDF417.Internal.PDF417 and the method determineDimensions.

**Reply by mickagau** [reply]
Thank you for your answer Michael.

Yes this is the method i was trying to play with but without success so far.

Finally by reading source code, i think that i have more to play with  ImageAspectRatio, cols min max, rows min max parameter and the width and height in pixel of the image. No need to touch source code.

As far as i understand, algorithm to define rows and cols is based on 1 pixel for module width and a preferred ratio (image aspect ratio).
Then in rendering method, it calculates a scale factor according to width and height parameters in pixel.
So by playing with image aspect ratio and min max cols/rows, i should get the final rendering expected of 5 pixels for the width of a module according to my resolution.

Regards
Mickael

**Reply by mickagau** [reply]
Hello

Finally, it is not clear to me how is calculated the rows and cols quantity for pdf417
Does someone know what are the 2 const below and how they are calculated?
private const float DEFAULT_MODULE_WIDTH = 0.357f; //1px in mm
private const float HEIGHT = 2.0f; //mm

They are used to determine the cols and rows quantity for pdf417 in this formula.
float newRatio = (((float)BarcodeMatrix.COLUMN_WIDTH * cols + start_stop_width) * DEFAULT_MODULE_WIDTH) / (rows * HEIGHT);

I replaced the formula by
float newRatio = (((float)BarcodeMatrix.COLUMN_WIDTH * cols + start_stop_width) * DEFAULT_MODULE_WIDTH) / (rows * DEFAULT_MODULE_WIDTH * aspectRatio);

which for me makes more sense as we want to have a ratio in image based on rows and column (maybe i misunderstand the meaning of this parameter).
It works or at least i can read back several pdf417 that i generate with this formula but i am really not sure if it can generate a side effect somewhere :(

May be someone already did some study on this ImagaAspectRatio parameter and could share result?

Thank you
Mickael

**Reply by mickagau** [reply]
Hello
Sorry for this spam, just to finish the discussion :)
Finally my first comment was the good  one.
I had to play with min max columns and min max rows but the point is that Min Max columns number does not count the Start Stop, right and left row indicator so my settings were not good.
By adjusting quantity of data to be encoded, the size of the barcode and those min max rows cols, i can set the width bar as expected to 5 pixels and set it readable by the device on the field.

Once again, thank you for this great library.

Regards
Mickael

**Reply by micjahn** [reply]
You are welcome. I think it isn't spam. Perhaps another user finds the information useful, too.

**Comment by Danielxmu** [comment]
Hello Michael

I am VB programmer as well and used ZXing to create GS1  DataMatrix  as picture at windows GUI in Excel doc.
According to below code, I can get the result with datamatrix content and pic, but it didn't comply with GS1 DM format, how can I fixed it? Hope get your help...

`Function Encode_To_DM_Code_To_File()
    Dim Writer As IBarcodeWriter, cont As String
    Dim DMCodeOptions As New DatamatrixEncodingOptions
    Set Writer = New BarcodeWriter
    cont = "(8010)338911AGSH33020-0101+(93)220407+(90)421298573+(91)AGSH33020-01+(92)01+(8011)22040011"
    Writer.Format = BarcodeFormat_DATA_MATRIX
    Set Writer.Options = DMCodeOptions
    DMCodeOptions.Height = 200
    DMCodeOptions.Width = 200
    DMCodeOptions.Margin = 5
    DMCodeOptions.SymbolShape = SymbolShapeHint_FORCE_SQUARE
    Writer.WriteToFile cont, ThisWorkbook.Path & "\test.png", ImageFileFormat_Png
End Function`

**Comment by micjahn** [comment]
ZXing doesn't support human-readable content style as you did ("(0810)33...+(93)...".
Please build your content string like it is described in the wiki
https://github.com/micjahn/ZXing.Net/wiki/Generating-GS1-DataMatrix-and-GS1-Code128-barcodes

**Comment by pintureti** [comment]
Hello from Spain! I’m using Zxing on vb.net, but I have a question. I want use qr gs1, is it possible? Thanks
